Improve test cases for the current requirements

r7kamura opened the pull request:
1. Remove test cases about foreign key change because this gem doesn't support it now
2. Avoid shared_examples for better debuggability (I know it's debatable but please let me make this change...)

Output from `--format documentation`:

```
GuardAgainstPhysicalDelete
  when child is physically deleted
    decrements counter
  when deleted_at child is physically deleted
    raises error (FAILED - 1)
  when removed_at child is physically deleted
    raises error (FAILED - 2)
  when deleted_at child is logically deleted
    decrements counter
  when removed_at child is logically deleted
    decrements counter
  when logically deleted deleted_at child is updated
    does not change counter
  when logically deleted removed_at child is updated
    does not change counter
  when logically deleted deleted_at child is logically undeleted
    increments counter
  when logically deleted removed_at child is logically undeleted
    increments counter
```

r7kamura commented:
In Rails 5.2, `child.destroy` doesn't raise error.
@uasi This is a bug, right?

r7kamura commented:
Do we need some change to pass test on Rails 5.2, like this?

```diff
diff --git a/lib/guard_against_physical_delete/base.rb b/lib/guard_against_physical_delete/base.rb
index 153528c..dc31a6a 100644
--- a/lib/guard_against_physical_delete/base.rb
+++ b/lib/guard_against_physical_delete/base.rb
@@ -45,6 +45,16 @@ module GuardAgainstPhysicalDelete
     end

     module InstanceMethods
+      if ::ActiveRecord.version >= ::Gem::Version.new('5.2')
+        def destroy
+          unless self.class.delete_permitted?
+            raise ::GuardAgainstPhysicalDelete::PhysicalDeleteError, self.class.name
+          end
+
+          super
+        end
+      end
+
       def hard_delete
         self.class.physical_delete { destroy }
       end
```

r7kamura commented:
oh, is this the right situation that `#destroy` is not guarded (but `#delete` is) ? Therefore, I should use `#delete` on my test?

r7kamura commented:
I decided to use `#destroy` in the test for now.
Please review 🙌

uasi commented:
😱  `#destroy` and `#destroy!` should be guarded also... This is indeed a bug in guard_against_physical_delete. 1.0.2 and below seems to be unaffected.

r7kamura commented:
> 😱 #destroy and #destroy! should be guarded also... This is indeed a bug in guard_against_physical_delete. 1.0.2 and below seems to be unaffected.

I see.

Added some more test cases (and removed some too verbose test cases) to avoid future problems.

<details>
<summary>Before</summary>

```
Logical
  it should behave like preventing physical delete
    #delete
      without #physical_delete block
        is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
      with #physical_delete block
        is expected not to raise Exception
        delete successufly with nesting block
        multi threading
          does not affect another thread
    #delete_all
      without #physical_delete block
        is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
      with #physical_delete block
        is expected not to raise Exception
    #hard_delete
      do delete record
    #soft_delete
      set timestamp to delete flag
      invokes after_save
      validates fields

RemovedAtLogical
  it should behave like preventing physical delete
    #delete
      without #physical_delete block
        is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
      with #physical_delete block
        is expected not to raise Exception
        delete successufly with nesting block
        multi threading
          does not affect another thread
    #delete_all
      without #physical_delete block
        is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
      with #physical_delete block
        is expected not to raise Exception
    #hard_delete
      do delete record
    #soft_delete
      set timestamp to delete flag
      invokes after_save
      validates fields

Physical
  #delete
    is expected not to raise Exception

ActiveRecord::Relation
  is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError

GuardAgainstPhysicalDelete
  deleted_at child
    it should behave like counter cache
      update counter_cache with soft_delete
      update counter_cache with hard_delete
      dont calc soft deleted record with updating
      calc revived record
      calc revived record with changing parent
  removed_at child
    it should behave like counter cache
      update counter_cache with soft_delete
      update counter_cache with hard_delete
      dont calc soft deleted record with updating
      calc revived record
      calc revived record with changing parent
  phisical delete
    update counter_cache with hard_delete

Finished in 0.86691 seconds (files took 0.24703 seconds to load)
33 examples, 0 failures

Coverage report generated for RSpec to /home/r7kamura/ghq/github.com/r7kamura/guard_against_physical_delete/coverage. 99 / 107 LOC (92.52%) covered.
```

</details>


<details>
<summary>After</summary>

```
GuardAgainstPhysicalDelete
  .physical_delete
    allows physical delete in given block
    with multi threading
      does not affect another thread
  .delete_all
    when model has no deleted_at column
      is expected not to raise Exception
    when model has deleted_at column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
    when model has logical_delete_column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
  .delete_all on relation
    when model has no deleted_at column
      is expected not to raise Exception
    when model has deleted_at column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
    when model has logical_delete_column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
  .destroy_all
    when model has no deleted_at column
      is expected not to raise Exception
    when model has deleted_at column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
    when model has logical_delete_column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
  .destroy_all on relation
    when model has no deleted_at column
      is expected not to raise Exception
    when model has deleted_at column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
    when model has logical_delete_column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
  #delete
    when model has no deleted_at column
      is expected not to raise Exception
    when model has deleted_at column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
    when model has logical_delete_column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
  #destroy
    when model has no deleted_at column
      is expected not to raise Exception
    when model has deleted_at column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
    when model has logical_delete_column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
  #destroy!
    when model has no deleted_at column
      is expected not to raise Exception
    when model has deleted_at column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
    when model has logical_delete_column
      is expected to raise GuardAgainstPhysicalDelete::PhysicalDeleteError
  #hard_delete
    destroys record in physical_delete block
  #soft_delete
    sets current time to logical delete column
    invokes after_save
    invokes validation

GuardAgainstPhysicalDelete
  when child is physically deleted
    decrements counter
  when deleted_at child is physically deleted
    raises error
  when removed_at child is physically deleted
    raises error
  when deleted_at child is logically deleted
    decrements counter
  when removed_at child is logically deleted
    decrements counter
  when logically deleted deleted_at child is updated
    does not change counter
  when logically deleted removed_at child is updated
    does not change counter
  when logically deleted deleted_at child is logically undeleted
    increments counter
  when logically deleted removed_at child is logically undeleted
    increments counter

Finished in 0.84083 seconds (files took 0.24749 seconds to load)
36 examples, 0 failures

Coverage report generated for RSpec to /home/r7kamura/ghq/github.com/r7kamura/guard_against_physical_delete/coverage. 99 / 107 LOC (92.52%) covered.
```

</details>

r7kamura commented:
All the commits I was trying to add in this pull request were merged at #30, so I'll close this. 👋
